### PR TITLE
[SPARK-41171][SQL] Push down filter through window when partitionSpec is empty

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Sometimes, the SQL exists filter which condition compares rank-like window functions with number. For example,
```
SELECT *,
         ROW_NUMBER() OVER(ORDER BY a) AS rn
FROM Tab1
WHERE rn <= 5
```
We can create a `Limit(5)` and push down it as the child of `Window`.
```
SELECT *,
         ROW_NUMBER() OVER(ORDER BY a) AS rn
FROM 
    (SELECT *
    FROM Tab1
    ORDER BY  a LIMIT 5) t
```

In short, it supports following pattern:
```
SELECT (... (row_number|rank|dense_rank)()
    OVER (
ORDER BY  ... ) AS rn)
WHERE rn (==|<|<=) k
        AND other conditions
```
For these three rank functions (row_number|rank|dense_rank), the rank of a key computed on dataset always <= its total rows of whole dataset，so we can safely discard rows with rank > k, anywhere.

This PR also take over some functions from https://github.com/apache/spark/pull/34367.

### Why are the changes needed?
1. reduce the shuffle write
2. solve skewed-window problem.
3. improve the performance and TPC-DS.

### Does this PR introduce _any_ user-facing change?
'No'.
Just update the inner implementation.

### How was this patch tested?
1. New test suites.
2. new micro benchmark
Before this PR
```
[info] Benchmark Top-K:                          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] ROW_NUMBER WITHOUT PARTITION                       10869          11498         928          1.9         518.3       1.0X
```
After this PR
```
[info] Benchmark Top-K:                          Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
[info] ------------------------------------------------------------------------------------------------------------------------
[info] ROW_NUMBER WITHOUT PARTITION                       1578           1718         191         13.3          75.3       1.0X
```
3. manual test on TPC-DS
**Micro Benchmark**
TPC-DS data size: 2TB.
I created a variety of q67 by remove the window partition specification.
image
This improvement is valid for the variety of tpcds q67 and no regression for other test cases.

| TPC-DS Query   | Before(Seconds)  | After(Seconds)  | Speedup(Percent)  |
|  ----  | ----  | ----  | ----  |
| variety of q67 | 3204.723 | 307.2745 | 942.95% |
| All TPC-DS | 9249.702 | 6373.544 | 45.13% |

### CODE REVIEW

**Comment by beliefer** [comment]
ping @zhengruifeng cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala [hunk 1]
```diff
@@ -17,19 +17,21 @@
...
-import org.apache.spark.sql.catalyst.trees.TreePattern.{LIMIT, WINDOW}
+import org.apache.spark.sql.catalyst.trees.TreePattern.{FILTER, LIMIT, WINDOW}
 
 /**
  * Pushes down [[LocalLimit]] beneath WINDOW. This rule optimizes the following case:
  * {{{
  *   SELECT *, ROW_NUMBER() OVER(ORDER BY a) AS rn FROM Tab1 LIMIT 5 ==>
  *   SELECT *, ROW_NUMBER() OVER(ORDER BY a) AS rn FROM (SELECT * FROM Tab1 ORDER BY a LIMIT 5) t
+ *   SELECT *, ROW_NUMBER() OVER(ORDER BY a) AS rn FROM Tab1 WHERE rn <= 5 ==>
+ *   SELECT *, ROW_NUMBER() OVER(ORDER BY a) AS rn FROM (SELECT * FROM Tab1 ORDER BY a LIMIT 5) t
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala:31]: Can we add a new rule? This isn't really limit push down, but more like infer limit.

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/LimitPushDownThroughWindow.scala:31]: OK

**Comment by beliefer** [comment]
https://github.com/apache/spark/pull/40142 replaces this one.

**Pull request closed without merging**